### PR TITLE
add BilinearInterpolationTest

### PR DESCRIPTION
**Pull request by lydonchandra**
Fixes # .

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [ ] I have updated API documentation
- [x] I have included unit tests
- [ ] I have updated the change log
- [ ] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Add tests for BilinearInterpolation RescaleGrayscale

### CODE REVIEW

File: Tests/Desktop/Imaging/Algorithms/BilinearInterpolationTest.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            Assert.Equal(
+                new byte[]
+                {
+                    0, 0, 0, 0, 0, 0, 0, 0,
+                    2, 2, 2, 2, 2, 2, 2, 2,
+                    5, 5, 5, 5, 5, 5, 5, 5,
+                    7, 7, 7, 7, 7, 7, 7, 7,
+                    10, 10, 10, 10, 10, 10, 10, 10,
+                    12, 12, 12, 12, 12, 12, 12, 12,
+                    15, 15, 15, 15, 15, 15, 15, 15,
```
Comment (lydonchandra) [inline Tests/Desktop/Imaging/Algorithms/BilinearInterpolationTest.cs:58]: This "second last row" is a duplicate of the last row. Could we confirm this is expected based on the algorithm ?

File: Tests/Desktop/Imaging/Algorithms/BilinearInterpolationTest.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            const int outputWidth = 4, outputHeight = 4;
+
+            byte[] output = BilinearInterpolation.RescaleGrayscale(input, inputWidth, inputHeight, outputWidth, outputHeight);
+
+            Assert.Equal(
+                new byte[]
+                {
+                    0, 0, 0, 0,
+                    5, 5, 5, 5,
+                    10, 10, 10, 10,
```
Comment (lydonchandra) [inline Tests/Desktop/Imaging/Algorithms/BilinearInterpolationTest.cs:25]: This "second last row" is a duplicate of the last row. Could we confirm this is expected based on the algorithm ?

**Comment by amoerie** [comment]
Hi, thanks for contributing! Could you elaborate a little on what made you create this pull request, and what you are trying to do here?

**Comment by lydonchandra** [comment]
Hi, didn't see any tests associated with the algorithms, which is why I added them

**Comment by gofal** [comment]
there can never be too much automated testing 😃

**Comment by gofal** [comment]
you forked this branch many commits ago on 12th July. some issues with flaky tests have been fixed since then. so the failing unit test on the v4-net-framework build may be of some other reason than your unit test. 
Could you rebase this branch or merge the last changes? Then the unit tests hould run green.

File: Tests/Desktop/Imaging/Algorithms/BilinearInterpolationTest.cs (hunk 1 above)
Comment (gofal) [reply Tests/Desktop/Imaging/Algorithms/BilinearInterpolationTest.cs:108]: Hm, not sure. My understanding of linear interpolation would say, that the last row shall not be repeated. I suppose you found a bug

**Comment by gofal** [comment]
I found the issue, actually the number of deltas have to be interpolated not the number of pixels. 

Previously the "scale-factor" was calculated by `inputWidth/outputWidth` and then for every output-pixel the coordinate was multiplied by this scale factor and then it was linear interpolated. But in fact we have to calcualte the relation of the number of Differences, so '(inputWidth-1)/(outputWidth-1)'. This value is very much the same if you take big numbers, but is very different if you take small meshes.

I also updated the expected values in the unit tests and now the linear interpolation is much more smooth!

**Comment by gofal** [comment]
One thing to notice: currently when it comes to rounding in the output matrix, fo-dicom does just a double-to-integer casting, which is always a floor rounding.  Is that OK? or should it be a Math.Round? math.Round would be more accurate but it also would be far more slower than just doing the integer casting.

What is your oppinion?

**Comment by lydonchandra** [comment]
Wrt. Math.Round vs integer casting, doubt it will make a lot of difference, let us do some benchmarking first :-)

**Comment by lydonchandra** [comment]
Also with unit tests in place, would like to attempt to vectorize the bilinear interpolation code for .NETCore as per 
https://fastcpp.blogspot.com/2011/06/bilinear-pixel-interpolation-using-sse.html
https://gist.github.com/lydonchandra/220eb2aeeae8a9542290a249349ba9a8
https://devblogs.microsoft.com/dotnet/hardware-intrinsics-in-net-core/

**Comment by lydonchandra** [comment]
And yes to be more accurate, some form for rounding other than flooring should be done. 

Math.Round(, MidpointRounding AwayFromZero) maybe ?

AwayFromZero | 1 | Round to nearest mode: when a number is halfway between two others, it is rounded toward the nearest number that is away from zero.

What do you think?

Comparison between floored output and raw output:
https://gist.github.com/lydonchandra/a00823466cff8b59ec880e4688288291

**Comment by lydonchandra** [comment]
You are correct @gofal

Double to integer casting is ~4x faster than Math.Round (mean of 1.3ms vs 5.3ms)

|                     Method                | InputSizePx | OutputSizePx |     Mean |     Error |    StdDev |
|---------------------------          |------------   |------------- |---------     :|----------:|----------:|
|       RescaleGrayscale_Ori         |         500     |         1000    | 1.346 ms   | 0.0234 ms | 0.0219 ms |
| RescaleGrayscale_MathRound |         500     |         1000    | 5.238 ms   | 0.1040 ms | 0.1388 ms |

**Comment by gofal** [comment]
if we assume only positive numbers, which should be true when dealing with pixel values, then we could simulate a Math.round by

    var rounded = (int)(value + 0.5);

adding a half and then doing double to integer casting. This should give a better performance. But as said, the requirement of only having positive values is important.

**Pull request closed without merging**